src/NBomber/Domain/Domain.fs

```diff
@@ -1,4 +1,4 @@
-﻿module internal rec NBomber.Domain
+﻿namespace NBomber.Domain
```

Comment:
Does it mean that Domain will be visible for users?
And what about **rec** ? what the benefit of removing it?

Reply:
Yes it means that `Domain` will be visible to users but that was not the intent. The intent here was to define a _namespace_ for all the domain realted code. So, to keep the Domain hidden from users, we can do something like:
```
namespace NBomber.Domain
module internal Domain =
.
.
.
```
Also later we can add **Errors.fs** to the **Domain** namespace. What do you think? 

#### Regarding removing rec
The intention is to force developer to be deliberate/intentional about dependencies at all levels of code. I think by using `rec` we are encouranging cyclic references and I think [Cyclic dependencies are evil](https://fsharpforfunandprofit.com/posts/cyclic-dependencies/) (You might also find [this comparison](https://fsharpforfunandprofit.com/posts/cycles-and-modularity-in-the-wild/) of C# and F# projects interesting). 
Also, I think `rec` is being overused here as noted in the **Drawbacks** section of the this [RFC](https://github.com/fsharp/fslang-design/blob/master/FSharp-4.1/FS-1009-mutually-referential-types-and-modules-single-scope.md):

>One risk of the proposal is that the feature will get over-used by beginners as the "easy" answer to avoid thinking about order of code. It may also be used by those looking to do top-down development (e.g. putting main first).

In my Opinion we should remove `rec` from where it's not necessary. Thoughts?

Reply:
@rohit-sharma 
'Yes it means that Domain will be visible to users but that was not the intent.'
This is the problem, we don't want to expose any of Domain's types.
Yes, we can define namespace Domain and then put hidden/internal modules with type definitions, but currently, it doesn't make any sense. 

'Also later we can add Errors.fs' - defining Domain as a namespace is making sense only if you gonna introduce new modules in separated files, like Step, TestFlow. As I mention before I don't see the benefit of splitting Domain into many files(very C#/OOP style), at least at this stage with current code size.
 
'Regarding removing rec'
We use rec by default, it helps you to move down the private method.  
Regarding - Cyclic dependencies are evil. As you may know, F# requires ordering of files, which is the real solver of this issue. We think that preventing rec is not a game changes.

Reply:
@AntyaDev Agree with not having an explicit namespace for domain.

Reply:
For not using **rec** modules, I am still convinced that it is better to not use **rec** where it is not required. I think it forces the programmer to think more about having cleaner dependency graph. 

Also, regarding - Cyclic dependencies are evil - it also mentions:
>You might say, “Yes, I can understand why circular dependencies are bad for assemblies, but why bother for code inside an assembly?”
>Well, for exactly the same reasons! Layering allows better partitioning, easier testing and cleaner refactoring.

and this...
>but my experience is that everything in the world is better when you’re forced to consider and manage “dependency order among software components” at every level of the system.

Reply:
@rohit-sharma I have updated my answer regarding rec

Reply:
@rohit-sharma I am not against reasoning about circular dependencies.
You might say, “Yes, I can understand why circular dependencies are bad for assemblies, but why to bother for code inside an assembly?”

F# order of files **is the real solver** of cyclic dependencies. At least what I have seen in practice.
Look at our project file structure as an example.

Reply:
@rohit-sharma 
>You can focus on the API and keep the private methods together by using nested functions. 
 Even C# has Nested/Local methods whose intent is:
I tend to disagree with this. 

1) In any way even if they are a local function the location will be the same, I mean that they will be at TOP. I don't need this, as I mentioned before, it's not the same as C# private/internal methods at the end of the file. It doesn't solve the issue with ordering from top to bottom. in Scala, for example, you don't have this limitation and everybody happy.
2) The local function will not be reused. The idea of local function is just a named procedure which will be used in one place. But it's not the same as I want from an internal function which could be reused in many places.
3) Local functions will not be testable. You just can't invoke them from outside. It's a big downside. 

>There's no need to re-order functions constantly, you just add a new function at the right place, that's basically it.
It happening during refactoring. officially I am working with 3 teams which use F# for about 2 years in a row. We use rec always by default. By the way, the rec feature was a gamechanger for my teams to start using F# because it gives you a freedom which you have in C#. It's not bad.

> Even the F# compiler uses module rec very scarcely
First of all, compilers it's a different kind of software, they have different tradeoffs. Second, you can find for example that we don't use carried functions by default, and what? Scala uses tuples for decades, and it doesn't make Scala less functional. We use carried args only if we need the partial application, it's our code style.

@rohit-sharma Regarding NBomber, it's not a main issue for us right now. We have a lot of much valuable tasks to do then just cosmetic changes which brings 0 value.

Reply:
Well, if you don't see any value in it, then I don't see the point of disucssing it any further. But thanks for taking the time to discuss 👍